### PR TITLE
feat(applications-service-api): get applications filtering

### PR DESCRIPTION
**Pull request by JamesP-KC**
amendments to the `GET /api/v1/applications` endpoint to allow filtering by stage, region and sector

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -62,7 +62,7 @@ services:
       - ./packages/common:/opt/app/packages/common
       - ./packages/applications-service-api:/opt/app/packages/applications-service-api
       - ./uploads:/opt/app/uploads
-    #    command: npm run start:dev:debug --workspace=packages/applications-service-api
+#    command: npm run start:dev:debug --workspace=packages/applications-service-api
```
Comment (Yakitrak) [inline docker-compose.yml:65]: Remember to revert :)

File: packages/applications-service-api/src/utils/application.mapper.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+		{ ni: 3, api: 'pre_examination', label: 'Pre-examination' },
+		{ ni: 4, api: 'examination', label: 'Examination' },
+		{ ni: 5, api: 'recommendation', label: 'Recommendation' },
+		{ ni: 6, api: 'decision', label: 'Decision' },
+		{ ni: 7, api: 'post_decision', label: 'Post-decision' },
+		{ ni: 8, api: 'withdrawn', label: 'Withdrawn' }
+	]
+};
+
+const mapFilterLabelToApi = (name, value) => {
```
Comment (Yakitrak) [inline packages/applications-service-api/src/utils/application.mapper.js:25]: Do you think its worth having unit tests for the smaller functions in this file?

File: packages/applications-service-api/src/utils/application.mapper.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+		{
+			stage: {},
+			region: {},
+			sector: {}
+		}
+	);
+
+	const filters = [];
+	for (let [field, filterValues] of Object.entries(mappedFilters)) {
+		for (let [value, count] of Object.entries(filterValues)) {
```
[the remark this answers is not in the document]
Comment (JamesP-KC) [reply packages/applications-service-api/src/utils/application.mapper.js:85]: yep that's a good shout. Definitely don't need `let` here

File: packages/applications-service-api/src/utils/application.mapper.js (hunk 2 above)
Comment (JamesP-KC) [reply packages/applications-service-api/src/utils/application.mapper.js:25]: I only test exported functions, and these ones are mostly there for a bit of extra readability. I will see about adding extra test cases to `buildApiFiltersFromNIApplications` and `mapApplicationFiltersToNI` though as there's only one at the moment

File: packages/applications-service-api/api/openapi.yaml [hunk 4]
```diff
@@ -85,6 +85,51 @@ paths:
               - +Stage
               - -Stage
               - Stage
+        - in: query
+          name: stage
```
[the remark this answers is not in the document]
Comment (JamesP-KC) [reply packages/applications-service-api/api/openapi.yaml:89]: as discussed, will leave it as-is for now - will be changing other query params / response attributes to camelcase in the near future so the sort params can be made camelcase then

File: packages/applications-service-api/__tests__/unit/services/application.service.test.js [hunk 5]
```diff
@@ -207,49 +216,33 @@ describe('application.service', () => {
...
-						LongLat: ['-0.70283147423378', '53.620078025496']
-					}
-				],
+				applications: APPLICATIONS_FO,
 				totalItems: mockCount,
 				itemsPerPage: 25,
 				totalPages: 4,
-				currentPage: 1
+				currentPage: 1,
+				filters: APPLICATIONS_FO_FILTERS
```
[the remark this answers is not in the document]
Comment (JamesP-KC) [reply packages/applications-service-api/__tests__/unit/services/application.service.test.js:224]: yep good idea. I've added a new `describe`

File: packages/applications-service-api/src/routes/applications.js [hunk 6]
```diff
@@ -21,6 +22,11 @@ const getApplicationsRoute = (req, res, next) => {
 
 router.get('/download', applicationsController.getAllApplicationsDownload);
 router.get('/:caseReference', validateRequestWithOpenAPI, getApplicationsRoute);
-router.get('/', applicationsController.getAllApplications);
+router.get(
+	'',
```
[the remark this answers is not in the document]
Comment (JamesP-KC) [reply packages/applications-service-api/src/routes/applications.js:29]: when you use a `/` here, the endpoint becomes `/api/v1/applications/` (trailing slash), which causes an issue when looking it up in the openapi spec (which doesn't have a trailing slash) to call the openapi validation function (https://github.com/Planning-Inspectorate/applications-service/blob/bb1f99c8f98a3318dce7d16f218b0cae1be7def5/packages/applications-service-api/src/utils/openapi.js#L25). Another way to fix that issue would be to strip the trailing `/` before looking up the path in the openapi spec

[the remark this answers is not in the document]
Comment (Yakitrak) [reply packages/applications-service-api/src/routes/applications.js:29]: Ah I see, makes sense - thank you

**Review by Yakitrak — APPROVED** [review]
LGTM

File: packages/applications-service-api/src/middleware/parseQueryParamProperties.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
+const { parseInteger } = require('../utils/parse');
```
Comment (Yakitrak) [inline packages/applications-service-api/src/middleware/parseQueryParamProperties.js:1]: Sorry approved but realised we should probably name this in kebab case like the other files

Comment (Yakitrak) [reply packages/applications-service-api/src/middleware/parseQueryParamProperties.js:1]: Applies to the corresponding test file too

**Pull request merged**